### ISSUE TITLE
Regression: Instantiating object having overloads in constructor call messes up inference of TypeVars.

### ISSUE THREAD

**Issue by GugelRobin** [microsoft/pyright#5021]
**Describe the bug**
Pyright 1.1.305 fails to correctly infer the generic type while 1.1.304 doesn't.

**To Reproduce**
* Check the code with pyright 1.1.304
* Observe: No issues found
* Check the code with 1.1.305
* Observe: `error: "assert_type" mismatch: expected "Bar[int]" but received "Bar[_T@Bar]` for last `assert_type` statement.

**Expected behavior**
Variable `c` (from code sample) should be inferred as `Bar[int]` instead of `Bar[_T@Bar]`

**Screenshots or Code**
```python
from typing import Generic, TypeVar, overload, Any
from typing_extensions import assert_type

_T = TypeVar("_T")

class Foo(Generic[_T]):
    @overload
    def __init__(self, _: _T):
        ...

    @overload
    def __init__(self, _: Any):
        ...

    def __init__(self, _: Any):
        ...

class Bar(Generic[_T]):

    def __init__(
        self,
        _: Foo[_T],
    ):
        ...

a = Foo(0)
b = Bar(a)
c = Bar(Foo(0))

assert_type(a, Foo[int])
assert_type(b, Bar[int])
assert_type(c, Bar[int]) # works with 1.1.304 - 1,1.305 raises the following error: 
# "assert_type" mismatch: expected "Bar[int]" but received "Bar[_T@Bar]" (reportGeneralTypeIssues)
```

**VS Code extension or command-line**
Executed via command line. Pylance ` Pylance language server 2023.4.40 (pyright 5b6a1d3b)` still infers the types correctly.
Python version 3.8

**Additional context**
For the Bug to occur `Foo` requires these overloads. The overloads are intentionally non-sensical to keep the sample small.
Possibly related to #4973

**Comment by erictraut** [comment]
This is addressed in pyright 1.1.308, which I just published. It will also be included in a future release of pylance.
